File: leetcode2025/2025_10/leetcode2025-10-06.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <queue>
#include <climits>
using namespace std;

class Solution {
public:
    int swimInWater(vector<vector<int>>& grid) {
        int n = grid.size();
        vector<vector<int>> dirs = {{0,1}, {0,-1}, {1,0}, {-1,0}};
        vector<vector<int>> dis(n, vector<int>(n, INT_MAX));
        dis[0][0] = 0;
        vector<vector<int>> visited(n, vector<int>(n, 0));
        priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, greater<tuple<int, int, int>>> pq;
        pq.emplace(0, 0, 0);
        while (!pq.empty()){
            auto [t, x, y] = pq.top();
            pq.pop();
            if (visited[x][y]){
                continue;
            }
            visited[x][y] = 1;
            for (int i=0; i<4; i++){
                int dx=dirs[i][0], dy=dirs[i][1];
                int nx=x+dx, ny=y+dy;
                if (nx<0 || nx>=n || ny<0 || ny>=n){
                    continue;
                }
                int cur = max(t, max(grid[x][y], grid[nx][ny]));
                if (cur < dis[nx][ny]){
                    dis[nx][ny] = cur;
                    pq.emplace(cur, nx, ny);
                }
            }
        }
        return dis[n-1][n-1];
    }
};

int main(){
    Solution s;
    vector<vector<int>> grid = {{0,2},{1,3}};
    cout << s.swimInWater(grid) << endl;
}

File: leetcode2025/2025_10/leetcode2025-10-06.py
from typing import List
from heapq import *

class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        n = len(grid)
        dirs = [(0,1), (0,-1), (1,0), (-1,0)] # 四个方向
        dis = [[float("inf")]*n for _ in range(n)] # 从(0,0)到各个位置的最短距离
        dis[0][0] = 0
        visited = [[0]*n for _ in range(n)] # 防止重复计算同一个点
        q = [(0, 0, 0)] # 小根堆，每一次都从当前已记录的距离最短的点开始向四周寻找
        while q:
            t, x, y = heappop(q)
            if visited[x][y]:
                continue
            visited[x][y] = 1
            for dx, dy in dirs:
                nx, ny = x+dx, y+dy
                if nx<0 or nx>=n or ny<0 or ny>=n:
                    continue
                cur = max(t, max(grid[x][y], grid[nx][ny])) # 到达(nx,ny)点的用时
                if cur < dis[nx][ny]:
                    dis[nx][ny] = cur
                    heappush(q, (cur, nx, ny))
        return dis[n-1][n-1]

if __name__ == "__main__":
    s = Solution()
    grid = [[0,2],[1,3]]
    print(s.swimInWater(grid))
